src/Core/src/Signer/SignerV4.php

```diff
@@ -101,6 +102,65 @@ protected function buildBodyDigest(Request $request, bool $isPresign): string
         return $hash;
     }
 
+    protected function convertBodyToStream(Request $request, \DateTimeImmutable $now, string $credentialString, string $signingKey, string &$signature): void
+    {
+        $body = $request->getBody();
+        if ($request->hasHeader('content-length')) {
+            $contentLength = (int) $request->getHeader('content-length');
+        } else {
+            $contentLength = $body->length();
+        }
+
+        // If content length is unknown, use the rewindable stream to read it once locally in order to get the length
+        if (null === $contentLength) {
+            $request->setBody($body = RewindableStream::create($body));
+            $body->read();
+            $contentLength = $body->length();
+        }
+
+        if (!$this instanceof SignerV4ForS3) {
+            $request->setBody($body = StringStream::create($body));
+
+            return;
+        }
+        // Code above is dedicated to S3 signature and has been moved to SignerV4ForS3.
+        // It can be removed in Core version 2.0
```

Comment:
Sorry for the confusion. 

All the code below is for S3 < 1.1, right?  So the "code above" is for every one. The code below is for old S3 clients. The code below should be removed.

Reply:
I mixed `above` and `below`.

So.
The code "below" works only for S3 (initial bug with SES)
The code "above" early exit for != S3 signer (everyone else). Fixe the bug
The code "below" keep compatibility with S3 < 1.1
(It can not be removed before 2.0 because S3: 1.0 requires Core 1.0)

is it more clear?

Reply:
Yes, it is better. :+1: